modules/apps/portal-security/portal-security-content-security-policy/src/main/java/com/liferay/portal/security/content/security/policy/internal/ContentSecurityPolicyHTMLRewriterImpl.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2024 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.portal.security.content.security.policy.internal;
+
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.content.security.policy.ContentSecurityPolicyHTMLRewriter;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.kernel.util.Validator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Attribute;
+import org.jsoup.nodes.Document;
+import org.jsoup.nodes.Element;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(service = ContentSecurityPolicyHTMLRewriter.class)
+public class ContentSecurityPolicyHTMLRewriterImpl
+	implements ContentSecurityPolicyHTMLRewriter {
+
+	@Override
+	public String rewriteInlineEventHandlers(String html, String nonce) {
+		Document document = Jsoup.parse(html);
+
+		Element body = document.body();
+
+		StringBundler sb = new StringBundler();
+
+		_extractInlineHandlers(body, sb);
+
+		if (sb.length() == 0) {
+			return html;
+		}
+
+		Element element = new Element("script");
+
+		element.attr("nonce", nonce);
+		element.html(sb.toString());
+
+		body.appendChild(element);
+
+		if (_containsBodyTag(html)) {
+			return body.outerHtml();
+		}
+
+		return body.html();
+	}
+
+	private boolean _containsBodyTag(String html) {
+		String trimmedHTML = html.trim();
+
+		String lowerCaseTrimmedHTML = StringUtil.toLowerCase(trimmedHTML);
+
+		return lowerCaseTrimmedHTML.startsWith("<body");
+	}
+
+	private void _extractInlineHandlers(Element element, StringBundler sb) {
+		String id = element.attr("id");
+
+		List<String> attributesToRemove = new ArrayList<>();
+
+		for (Attribute attribute : element.attributes()) {
+			String key = attribute.getKey();
+
+			String lowerCaseKey = StringUtil.toLowerCase(key);
+
+			if (lowerCaseKey.startsWith("on")) {
```

Comment:
What about invalid attrs? `onfoobar=""` that'd technically pass this, right? Maybe it's not a big deal though

Reply:
So I had two options:

1. Maintain a long list of `on...` handlers that may become incomplete in the future if new events appear and that may be missing others if they are deprecated (eg: `onkeypress` doesn't show up in the MDN because it was deprecated).
2. Assume each attribute starting with `on` is a handler.

I think 2 is better because it's more robust and if the event handler does not exist, the code won't fail. 

There's of course the possibility that someone is using a non standard `onwhatever` attribute to do fancy things with JavaScript and the tag breaks that, but:

1. I find it quite improbable.
2. People shouldn't invent `on...` handlers because they can collide in the future.

So I assume we are safe. 

What do you think?

Reply:
Agreed to leave this as is.